src/libraries/Native/Unix/System.Native/pal_signal.c

```diff
@@ -17,33 +17,179 @@
+            *posixSignal = PosixSignalSIGCONT;
+            return true;
+
+        case SIGTTIN:
+            *posixSignal = PosixSignalSIGTTIN;
+            return true;
+
+        case SIGTTOU:
+            *posixSignal = PosixSignalSIGTTOU;
+            return true;
+
+        case SIGTSTP:
+            *posixSignal = PosixSignalSIGTSTP;
+            return true;
+
+        default:
+            *posixSignal = signalCode;
+            return false;
+    }
+}
+
+int32_t SystemNative_GetPlatformSignalNumber(PosixSignal signal)
 {
-    switch (sig)
+    switch (signal)
+    {
+        case PosixSignalSIGHUP:
+            return SIGHUP;
+
+        case PosixSignalSIGINT:
+            return SIGINT;
+
+        case PosixSignalSIGQUIT:
+            return SIGQUIT;
+
+        case PosixSignalSIGTERM:
+            return SIGTERM;
+
+        case PosixSignalSIGCHLD:
+            return SIGCHLD;
+
+        case PosixSignalSIGWINCH:
+            return SIGWINCH;
+
+        case PosixSignalSIGCONT:
+            return SIGCONT;
+
+        case PosixSignalSIGTTIN:
+            return SIGTTIN;
+
+        case PosixSignalSIGTTOU:
+            return SIGTTOU;
+
+        case PosixSignalSIGTSTP:
+            return SIGTSTP;
+    }
+
+    if (   signal > 0
+        && signal <= NSIG      // Ensure we stay within the static arrays.
+#ifdef SIGRTMAX
+        && signal <= SIGRTMAX  // Runtime check for highest value.
+#endif
+       )
     {
-        case SIGINT:  return &g_origSigIntHandler;
-        case SIGQUIT: return &g_origSigQuitHandler;
-        case SIGCONT: return &g_origSigContHandler;
-        case SIGCHLD: return &g_origSigChldHandler;
-        case SIGWINCH: return &g_origSigWinchHandler;
+        return signal;
     }
 
-    assert(false);
-    return NULL;
+    return 0;
+}
+
+void SystemNative_SetPosixSignalHandler(PosixSignalHandler signalHandler)
+{
+    assert(signalHandler);
+    assert(g_posixSignalHandler == NULL || g_posixSignalHandler == signalHandler);
+
+    g_posixSignalHandler = signalHandler;
+}
+
+static struct sigaction* OrigActionFor(int sig)
+{
+    return &g_origSigHandler[sig - 1];
+}
+
+static void RestoreSignalHandler(int sig)
+{
+    bool* isInstalled = &g_handlerIsInstalled[sig - 1];
+    *isInstalled = false;
+    sigaction(sig, OrigActionFor(sig), NULL);
 }
 
 static void SignalHandler(int sig, siginfo_t* siginfo, void* context)
 {
-    // Signal handler for signals where we want our background thread to do the real processing.
-    // It simply writes the signal code to a pipe that's read by the thread.
+    if (sig == SIGCONT)
+    {
+        ConsoleSigTtouHandler consoleTtouHandler = g_consoleTtouHandler;
+        if (consoleTtouHandler != NULL)
+        {
+            consoleTtouHandler();
+        }
+    }
+
+    // Signals that can be canceled through the PosixSignal API are handled later.
+    // For other signals, we immediately invoke the original handler.
+    if (sig != SIGQUIT &&
+        sig != SIGTERM &&
```

Comment:
I might be misremembering, but doesn't CoreCLR install handlers for some of these before System.Native? Wouldn't that effectively render them uncancelable?

Reply:
That's right. The default behavior is assuming original signal handlers must be called. Handlers are cancelable only if specific support is added for it.